### DISCUSSION TITLE
How could should I create a PrismaAdapter

### DISCUSSION THREAD

**Discussion by Noudea** [prisma/prisma#17289] [Q&A]
I would like to create a PrismaAdapter,the model should be a string or the PrismaModel (i dont know how to type it) but I am struggling and I don't know if it is even possible. Here is my failed attempt.

thx for your help

```ts

import { PrismaClient } from '@prisma/client'



//type PrismaModel = keyof PrismaClient<PrismaClientOptions>
//type PrismaModel = keyof Prisma.userDelegate<GlobalReject>
//type PrismaUserModel = Prisma.userDelegate<GlobalReject>;


class PrismaAdapter {
  private prisma: PrismaClient
  private user: PrismaUserModel

  constructor() {
    this.prisma = new PrismaClient()
    this.user = this.prisma.user
  }

  async findOne(model: PrismaModel, where: object): Promise<any> {
    return await this.prisma[model].findOne({ where })
  }

  async findMany(model: string): Promise<any[]> {
    return await this.prisma[model].findMany()
  }

  async create(model: PrismaModel, data: object): Promise<any> {
    return await this.prisma[model].create({ data })
  }

  async update(model: string, where: object, data: object): Promise<any> {
    return await this.prisma[model].update({ where, data })
  }

  async delete(model: string, where: object): Promise<any> {
    return await this.prisma[model].delete({ where })
  }
}

export default PrismaAdapter



```

I am expecting to use it in a Database service class.

```ts
import PrismaAdapter from "./PrismaAdapter";


class DatabaseAdapter {
  private database: PrismaAdapter;
  private model: PrismaModel;

  constructor({database, model}: {database: PrismaAdapter, model: PrismaModel}) {
    this.database = database;
  }

  async findOne(model :PrismaModel,id: number): Promise<any> {
    return this.database.findOne(model,where: {id})
  }
  
  async findMany(model: string): Promise<any[]> {
    return await this.database.findMany(model)
  }

}

export default DatabaseAdapter

```

And use this database Adapter in for exemple a UserRepository.

**Comment by nurul3101** [comment]
Hi @Noudea 👋 

As a starting point, I would recommend you to take this official [example](https://github.com/prisma/prisma-examples/tree/latest/typescript/rest-nestjs) as a reference. This example has a PrismaService that is used in other controllers. This example is minimal but should be helpful.

**Reply by Noudea** [reply]
thx, will look into it

**Comment by Noudea** [comment]
So this is the solution I used :

Just need to type the return and add a few methods.

Will update when finished

```ts


class DatabaseAdapter {
  private database: any;
  private model: string;

  constructor({database, model}: {database: any, model: string}) {
    this.database = database;
    this.model = model;
  }

  async create({data,select,include}: { data:object,select?: object,include?: object }): Promise<any> {
    return await this.database[this.model].create({data})
  }

  async findOne({where,select,include}: {where: object,select?: object,include?: object}): Promise<any> {
    return await this.database[this.model].findUnique({where,select,include})
  }

  async update({where,data}: {where: object,data: object}): Promise<any> {
    return await this.database[this.model].update({where,data})
  }


}
//TODO implement the rest

export default DatabaseAdapter



```

**Reply by nurul3101** [reply]
Thanks for sharing the solution! 👏

**Reply by Noudea** [reply]
thx I will update my solutions when I'm finished 100%

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
